File: problems/42/solution.py
# Accepted: 100ms (91.51%), 19.26MB (21.43%)

class Solution:
    def trap(self, height: List[int]) -> int:
        maxHeight = max(height)
        maxHeightIndex = height.index(maxHeight)

        water = 0

        # from left to the peak
        maxHeightSoFar = 0
        for i in range(maxHeightIndex):
            maxHeightSoFar = max(maxHeightSoFar, height[i])
            water += maxHeightSoFar - height[i]

        # from right to the peak
        maxHeightSoFar = 0
        for i in reversed(range(maxHeightIndex+1, len(height))):
            maxHeightSoFar = max(maxHeightSoFar, height[i])
            water += maxHeightSoFar - height[i]

        return water

File: problems/42/solution.ts
// Accepted: 56ms (88.93%), 45.01MB (83.75%)

function trap(height: number[]): number {
    const maxHeight = Math.max(...height);
    const maxHeightIndex = height.indexOf(maxHeight);

    let water = 0;

    // from left to the peak
    let maxHeightSoFar = 0;
    for (let i = 0; i < maxHeightIndex; ++i) {
        maxHeightSoFar = Math.max(maxHeightSoFar, height[i]);
        water += maxHeightSoFar - height[i];
    }

    // from right to the peak
    maxHeightSoFar = 0;
    for (let i = height.length-1; i > maxHeightIndex; --i) {
        maxHeightSoFar = Math.max(maxHeightSoFar, height[i]);
        water += maxHeightSoFar - height[i];
    }

    return water;
};
